applications/imaging_ai_segmentator/monai_totalseg_operator.py

```diff
@@ -0,0 +1,283 @@
+
+    def _find_model_file_path(self, model_path: Path):
+        # We'd just have to find the first file and return it for now,
+        # because it is known that in Holoscan App Package, the path is the root models folder
+        if model_path:
+            if model_path.is_file():
+                return model_path
+            elif model_path.is_dir():
+                for file in model_path.rglob("*"):
+                    if file.is_file():
+                        return file
+
+        raise ValueError(f"Model file not found in the provided path: {model_path}")
+
+    def setup(self, spec: OperatorSpec):
+        spec.input(self.input_name_image)
+        spec.output(self.output_name_seg)
+        spec.output(self.output_name_saved_images_folder).condition(
+            ConditionType.NONE
+        )  # Output not requiring a receiver
+
+    def compute(self, op_input, op_output, context):
+        input_image = op_input.receive(self.input_name_image)
+        if not input_image:
+            raise ValueError("Input image is not found.")
+
+        # This operator gets an in-memory Image object, so a specialized ImageReader is needed.
+        _reader = InMemImageReader(input_image)
+
+        # In this example, the input image, once loaded at the beginning of the pre-transforms, can
+        # be saved on disk, so can the segmentation prediction image at the end of the post-transform.
+        # They are both saved in the same subfolder of the application output folder, with names
+        # distinguished by the postfix. They can also be saved in different subfolder if need be.
+        # These images files can then be packaged for rendering.
+        # In the code below, saving of the image files are disabled to save 10 seconds if nii, and 20 if nii.gz
+        pre_transforms = self.pre_process(_reader, str(self.output_folder))
+        post_transforms = self.post_process(pre_transforms, str(self.output_folder))
+
+        # Load the PyTorch model directly as the model factory limits the support to TorchScript as of now
+
+        _device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        _blocks_down: tuple = (1, 2, 2, 4)
+        _blocks_up: tuple = (1, 1, 1)
+
+        # Create the model and push to device
+        model = monai.networks.nets.segresnet.SegResNet(
+            spatial_dims=3,
+            in_channels=1,
+            out_channels=105,
+            init_filters=32,
+            blocks_down=_blocks_down,
+            blocks_up=_blocks_up,
+            dropout_prob=0.2,
+        ).to(_device)
+
+        model.load_state_dict(torch.load(self.model_path))
+        model.eval()
+
+        loaded_model = Model(self.model_path, name="monai_total_seg")
+        loaded_model.predictor = model
+
+        self.app_context.models = loaded_model
+        # Delegates inference and saving output to the built-in operator.
+        infer_operator = MonaiSegInferenceOperator(
```

Comment:
Same question as above - can we move this out of `compute` to reduce initialization cost?

Seems a little unusual to have an operator member that is not connected to a workflow. I wonder if this would benefit from refactoring to use a common, non-operator impl class for reuse as a member in both operators.

Reply:
Original use case and hence design of the application is for handling container workload, i.e. discrete payload and short lived container just for that workload, so it does not matter the model loading is done at init time or compute time. The container is only launched when a piece of workload is ready, different from typical Holoscan apps. Of course, when this app is made to handle continuous workload (streaming), impl has to take into consideration of init cost.
Another reason is that object instance and model parameters may need to be adjusted based on the specific input, though in this one, there may not seem to be.